HPC-8026:Allow selecting monitoring periods to include in plan publishing

krishnareddy-tadi opened the pull request.

krishnareddy-tadi committed:
Create plantag related resolvers, types and service files

krishnareddy-tadi committed:
Add plan library methods to update tags for entities

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on src/domain-services/plantag/plantag-service.ts:
```diff
@@ -0,0 +1,79 @@
+    });
+    if (!planTags || !planTags.length) {
+      throw new Error(`Plantags doesn't exist for planId ${planId}`);
+    }
```
Let's not throw an error here and instead return empty array
```suggestion
```

Pl217 commented on src/domain-services/plantag/plantag-service.ts:
```diff
@@ -0,0 +1,79 @@
+    models: Database,
+    planTag: AddPlanTagInput
+  ): Promise<InstanceDataOfModel<Database['planTag']>> {
+    console.log('hi', planTag.planId);
```
Hi :)
```suggestion
```

Pl217 commented on src/domain-services/plantag/graphql/resolver.ts:
```diff
@@ -0,0 +1,37 @@
+    @Ctx() context: Context
+  ): Promise<InstanceDataOfModel<Database['planTag']>[]> {
+    return await this.planTagService.findByPlanId(context.models, planId);
+  }
```
This should not be a `Query`, but `FieldResolver`. Look at how plan years are fetched from plan. For example, when we search for a plan, we can get its years, but we should expand to allow for fetching plan tags like this
```
{
  plan(id: 1061) {
    years,
    tags # ADD THIS
  }
}
```
All tables that derive from some base table like `planEmergency`, `planLocation`, `planTag`, etc. should go as field resolvers of a plan. Same goes for other tables e.g. `projectContact`, `budgetSegmentBreakdown`, etc.

Pl217 commented on src/domain-services/plantag/graphql/resolver.ts:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  @Mutation(() => PlanTag)
+  async publishplan(
```
This probably belongs logically under plan resolvers. That would only leave fetching by plan tag ID in this method, which I don't know how much it is usable, but let's keep that option to search plan tags by ID.

I also noticed that you didn't use camelCase in many cases, like methods in this file. Did you find it somewhere that GraphQL convention is against the usage of camelCase?

As for the directory that this file is in (`src/domain-services/plantag/graphql`), we already established usage of kebab-case, which means `plantag` -> `plan-tag`

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+    (latestPlanTag.length && latestPlanTag[0].name) || '0.0';
+  let [majorVersion, minorVersion] = latestTagName
+    .split('.')
+    .map((v) => parseInt(v) || 0);
```
Let's use a radix, just to be safe
```suggestion
    .map((v) => parseInt(v, 10) || 0);
```

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+import { Brand, createBrandedValue } from '@unocha/hpc-api-core/src/util/types';
+import { AddPlanTagInput } from '../../domain-services/plantag/graphql/types';
+
+export const getVersion = async (
```
`getVersion` is not a good name here. Plans have versions (tracked in `planVersion`), and those `planVersions` are tagged, using `versionTags` column. But also, plans are tagged (in `planTag` column). Plan's tags contain all the tags, whereas `planVersion`'s `versionTags` column usually contains only a subset of these tags.

I think `getNextTag` is the name which better describes what happens. `getVersion` implies that you fetch `planVersion` of a `plan`.

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+    where: {
+      planId: createBrandedValue(planTag.planId),
+    },
+    orderBy: { column: 'createdAt', order: 'desc' },
```
Is this how old code used to work? I think that latest plan tag (by `createdAt` column) is also the highest tag version, by its `name` field. If we order by `name` here, we can get the same result. But, don't feel pressured to make this change, do it only if you see it as a semantic improvement. We would rely on PostgreSQL string ordering algorithm to put the latest tag highest in order, but we can also fetch unordered set of plan tags and find highest tag ourselves.

krishnareddy-tadi replied:
There is no gurantee that we will get latest version always if we sort by name. Suppose we have version 1.0,1.1,1.2,1.3.........1.11,1.12,1.3 If sort based on the name we will get result like this 
1.0,1.1,1.11,1.12,1.13,1.2,1.3..... .

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+    orderBy: { column: 'periodNumber', order: 'desc' },
+  });
+
+  if (reportingPeriods.length) {
```
You can reduce indentation here
```suggestion
  if (!reportingPeriods.length) {
    return;
  }
```

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+        versionTags: { [models.Op.NOT_IN]: [['{}']] },
+      },
+    });
+    console.log(publishedMeasurements.length);
```
```suggestion
```

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+        models,
+        planTag
+      );
+      console.log(associatedPlanVersion);
```
```suggestion
```

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+  planReportingPeriodIds: PlanReportingPeriodId[] | []
+) => {
+  for (const versionedModel of versionedModels) {
+    console.log(versionedModel.table, versionedModel?.relationship);
```
```suggestion
```

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>,
+  publishMeasurements: boolean,
+  planReportingPeriodIds: PlanReportingPeriodId[] | []
```
What does this mean?
```suggestion
  planReportingPeriodIds: PlanReportingPeriodId[]
```

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+    'createdAt' | 'planId' | 'public' | 'name'
+  >
+) => {
+  const model = models._tables[tableName];
```
No need to access "private" field `_tables`, which is exposed for testing purposes.
```suggestion
  const model = models[tableName];
```
Replace other similar usages in this file

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+  const model = models._tables[tableName];
+  const versionModel = models._tables[`${tableName}Version`];
+  const idField = `${tableName}Id`;
+  type AnyModelId = Brand<number, any, any>;
```
Move this type definition to `src/common-utils/branding.ts` and import it from there.

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+  >
+) => {
+  const model = models._tables[tableName];
+  const idField = `planId`;
```
This doesn't make sense. Just use `planId` directly and not via computed property.
```suggestion
```

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+
+  const versionRows = await model.find({
+    where: {
+      [idField]: tag.planId,
```
```suggestion
      planId: tag.planId,
```

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+      id: {
+        [models.Op.NOT_IN]: latestVersionIds.map(createBrandedValue),
+      },
+      [idField]: tag.planId,
```
```suggestion
      planId: tag.planId,
```

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+  const versionRowsByPlanId = new Map<
+    unknown,
+    Set<InstanceDataOfModel<typeof model>>
+  >();
```
`import type { PlanId } from '@unocha/hpc-api-core/src/db/models/plan';`
```suggestion
  const versionRowsByPlanId = new Map<
    PlanId,
    Set<InstanceDataOfModel<typeof model>>
  >();
```

Pl217 commented on src/domain-services/plantag/graphql/types.ts:
```diff
@@ -0,0 +1,78 @@
+  registerEnumType,
+  InputType,
+} from 'type-graphql';
+import { Brand } from '../../../common-utils/branding';
```
Import `Brand` from `hpc-api-core`. We should delete `Brand` and `createBrandedValue` from `src/common-utils/branding.ts`

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+  }
+};
+
+const versionedModels: any = [
```
Don't use `any`, except in special circumstances when defining type generics.

From this point forward, all the code should be moved to a separate file, because it is all related to versioning, so that way we can have everything isolated in a single place. Read my main comment to see the proposed contents of this file.

Pl217 commented on src/common-libs/plan/index.ts:
```diff
@@ -0,0 +1,511 @@
+    if (versionedModel?.relationship) {
+      await updateRelationshipModelTags(models, versionedModel.table, planTag);
+    } else {
+      await updateEntityVersionModelTags(models, versionedModel.table, planTag);
```
This applies to `planYear` and `planLocation`, but these are "relationship" tables in old API codebase.

Pl217 reviewed (changes requested):
I have tried again to unify some common methods for versioning, but reached "Expression produces a union type that is too complex to represent" once again. So, I ended up having three methods related to three cases that we have for versioning:
1) Dealing with tables which require only base-level changes. This category is only for `planLocation` and `planYear` tables
2) Dealing with tables which require only version-level changes. This category only includes `planVersion` table.
3) Dealing with tables which require both base-level and version-level changes. This applies to all remaining tables.

With these three methods, we have some duplication still, but type safety is preserved and `yarn check-types` takes 6-7 seconds, which is acceptable (although higher than current ~4 we have with your code). With more complex types, my VS Code operates at sub-glacial speeds and checking types takes up to a whole minute.

I don't see why you treated `measurement` table separately, when it is only a special case in a way that data is fetched, as it goes via `attachment` table (and its `planId`) and not via own `planId` column, like in other cases.

There are some comments inline (although some no longer apply if you use my code below), but I suggest that you extract versioning code into a separate file (`src/common-libs/plan/versioning.ts`) and have something like this:
```ts
/**
 * This file contains the code that updates various tables
 * when a new plan version is published and plan gets tagged.
 *
 * It updates the following tables:
 * - attachment
 * - attachmentVersion
 * - governingEntity
 * - governingEntityVersion
 * - measurement
 * - measurementVersion
 * - planEntity
 * - planEntityVersion
 * - planLocation
 * - planVersion
 * - planYear
 *
 * This can basically be achieved with two types of methods,
 * one that updates base tables (like `attachment`) and the
 * other that updates version tables (like `attachmentVersion`).
 *
 * However, trying to optimize the code without duplication
 * leads to great TS slowdown and sometimes also triggering error:
 * "Expression produces a union type that is too complex to represent"
 *
 * In order to preserve type safety while working with models, at an
 * expense of some code duplication, three separate methods are used:
 * 1) Dealing with tables which require only base-level changes
 * 2) Dealing with tables which require only version-level changes. This
 * category only includes `planVersion` table.
 * 3) Dealing with tables which require both base-level and version-level changes
 */

import type { Database } from '@unocha/hpc-api-core/src/db';
import type { PlanReportingPeriodId } from '@unocha/hpc-api-core/src/db/models/planReportingPeriod';
import { InstanceDataOfModel } from '@unocha/hpc-api-core/src/db/util/raw-model';
import { groupObjectsByProperty } from '@unocha/hpc-api-core/src/util';
import { createBrandedValue } from '@unocha/hpc-api-core/src/util/types';
import type { AnyModelId } from '../../common-utils/branding';

const baseOnlyModels = ['planLocation', 'planYear'] as const;
const baseAndVersionModels = [
  'attachment',
  'governingEntity',
  'planEntity',
] as const;

type BaseOnlyModels = typeof baseOnlyModels[number];
type BaseAndVersionModels = typeof baseAndVersionModels[number] | 'measurement';

type IdType<T extends BaseAndVersionModels> = `${T}Id`;

const noLongerParanoidTables = new Set();
// TODO: Uncomment for HPC-8305
// const noLongerParanoidTables = new Set(['planYear']);

export const updateVersionStates = async (
  models: Database,
  planTag: InstanceDataOfModel<Database['planTag']>,
  publishMeasurements: boolean,
  planReportingPeriodIds: PlanReportingPeriodId[]
) => {
  for (const tableName of baseOnlyModels) {
    await updateBaseModelTags(models, tableName, planTag);
  }
  for (const tableName of baseAndVersionModels) {
    const baseRows = await models[tableName].find({
      where: {
        planId: planTag.planId,
      },
    });

    await updateBaseAndVersionModelTags(models, tableName, baseRows, planTag);
  }

  await updatePlanVersionTags(models, planTag);

  if (publishMeasurements) {
    const attachments = await models.attachment.find({
      where: {
        planId: planTag.planId,
      },
    });
    const measurements = await models.measurement.find({
      where: {
        attachmentId: {
          [models.Op.IN]: attachments.map((a) => a.id),
        },
        planReportingPeriodId: {
          [models.Op.IN]: planReportingPeriodIds,
        },
      },
    });

    await updateBaseAndVersionModelTags(
      models,
      'measurement',
      measurements,
      planTag
    );
  }
};

const updateBaseAndVersionModelTags = async (
  models: Database,
  tableName: BaseAndVersionModels,
  baseRows: InstanceDataOfModel<Database[BaseAndVersionModels]>[],
  tag: Pick<
    InstanceDataOfModel<Database['planTag']>,
    'createdAt' | 'planId' | 'public' | 'name'
  >
) => {
  const model = models[tableName];
  const versionModel = models[`${tableName}Version`];
  const idField = `${tableName}Id` as IdType<BaseAndVersionModels>;

  const inactiveRows: AnyModelId[] = [];
  for (const baseRow of baseRows) {
    const isActive =
      baseRow.deletedAt === null || baseRow.deletedAt > tag.createdAt;

    if (isActive) {
      await model.update({
        values: {
          latestTaggedVersion: true,
          versionTags: [...baseRow.versionTags, tag.name],
          ...(tag.public ? { currentVersion: true } : {}),
        },
        where: { id: createBrandedValue(baseRow.id) },
      });
    } else {
      inactiveRows.push(createBrandedValue(baseRow.id));
    }
  }

  await model.update({
    values: {
      latestTaggedVersion: false,
      ...(tag.public ? { currentVersion: false } : {}),
    },
    where: {
      id: { [models.Op.IN]: inactiveRows.map(createBrandedValue) },
    },
  });

  const versionRows = await versionModel.find({
    where: {
      [idField]: {
        [models.Op.IN]: baseRows.map((a) => a.id),
      },
      createdAt: { [models.Op.LTE]: tag.createdAt },
    },
  });

  const versionRowsByBaseId = new Map<
    unknown,
    Set<InstanceDataOfModel<typeof versionModel>>
  >();
  for (const obj of versionRows) {
    const value = obj[idField as keyof typeof obj];
    let group = versionRowsByBaseId.get(value);
    if (!group) {
      versionRowsByBaseId.set(value, (group = new Set()));
    }
    group.add(obj);
  }

  const latestVersionIds = [...versionRowsByBaseId.values()].map((versions) =>
    // Get biggest version ID
    Math.max(...[...versions].map((av) => av.id))
  );

  const latestVersions = await versionModel.find({
    where: {
      id: {
        [models.Op.IN]: latestVersionIds.map(createBrandedValue),
      },
    },
  });

  for (const latestVersion of latestVersions) {
    await versionModel.update({
      values: {
        latestTaggedVersion: true,
        versionTags: [
          ...latestVersion.versionTags,
          ...(latestVersion.versionTags.includes(tag.name) ? [] : [tag.name]),
        ],
        ...(tag.public ? { currentVersion: true } : {}),
      },
      where: { id: createBrandedValue(latestVersion.id) },
    });
  }

  const oldVersions = await versionModel.find({
    where: {
      id: {
        [models.Op.NOT_IN]: latestVersionIds.map(createBrandedValue),
      },
      [idField]: { [models.Op.IN]: baseRows.map((a) => a.id) },
    },
  });
  await versionModel.update({
    values: {
      latestTaggedVersion: false,
      ...(tag.public ? { currentVersion: false } : {}),
    },
    where: {
      id: {
        [models.Op.IN]: oldVersions.map((v) => v.id as AnyModelId),
      },
    },
  });
};

const updateBaseModelTags = async (
  models: Database,
  tableName: BaseOnlyModels,
  tag: Pick<
    InstanceDataOfModel<Database['planTag']>,
    'createdAt' | 'planId' | 'public' | 'name'
  >
) => {
  const model = models[tableName];

  const baseRows = await model.find({
    where: {
      planId: tag.planId,
    },
  });

  const inactiveRows: AnyModelId[] = [];
  for (const baseRow of baseRows) {
    const isActive =
      /**
       * If table is no longer paranoid (i.e. `deletedAt` column is
       * dropped from DB schema), just simply treat all records as active.
       */
      noLongerParanoidTables.has(tableName) ||
      baseRow.deletedAt === null ||
      baseRow.deletedAt > tag.createdAt;

    if (isActive) {
      await model.update({
        values: {
          latestTaggedVersion: true,
          versionTags: [...baseRow.versionTags, tag.name],
          ...(tag.public ? { currentVersion: true } : {}),
        },
        where: { id: createBrandedValue(baseRow.id) },
      });
    } else {
      inactiveRows.push(createBrandedValue(baseRow.id));
    }
  }

  await model.update({
    values: {
      latestTaggedVersion: false,
      ...(tag.public ? { currentVersion: false } : {}),
    },
    where: {
      id: { [models.Op.IN]: inactiveRows.map(createBrandedValue) },
    },
  });
};

const updatePlanVersionTags = async (
  models: Database,
  tag: Pick<
    InstanceDataOfModel<Database['planTag']>,
    'createdAt' | 'planId' | 'public' | 'name'
  >
) => {
  const planVersionsByPlanId = groupObjectsByProperty(
    await models.planVersion.find({
      where: {
        planId: tag.planId,
        createdAt: { [models.Op.LTE]: tag.createdAt },
      },
    }),
    'planId'
  );

  const latestPlanVersionIds = [...planVersionsByPlanId.values()].map(
    (planVersions) =>
      // Get biggest plan version ID
      Math.max(...[...planVersions].map((av) => av.id))
  );

  const latestVersions = await models.planVersion.find({
    where: {
      id: {
        [models.Op.IN]: latestPlanVersionIds.map(createBrandedValue),
      },
    },
  });

  for (const latestVersion of latestVersions) {
    await models.planVersion.update({
      values: {
        latestTaggedVersion: true,
        versionTags: [
          ...latestVersion.versionTags,
          ...(latestVersion.versionTags.includes(tag.name) ? [] : [tag.name]),
        ],
        ...(tag.public ? { currentVersion: true } : {}),
      },
      where: { id: latestVersion.id },
    });
  }

  const oldVersions = await models.planVersion.find({
    where: {
      id: {
        [models.Op.NOT_IN]: latestPlanVersionIds.map(createBrandedValue),
      },
      planId: { [models.Op.IN]: planVersionsByPlanId.keys() },
    },
  });
  await models.planVersion.update({
    values: {
      latestTaggedVersion: false,
      ...(tag.public ? { currentVersion: false } : {}),
    },
    where: {
      id: {
        [models.Op.IN]: oldVersions.map((v) => v.id as AnyModelId),
      },
    },
  });
};

```

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 committed:
⬆️ Update `@unocha/hpc-api-core` to `v4.4.4`

Pl217 committed:
HPC-8305: Exclude `planYear` from paranoid tables

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 reviewed (approved):
This is yet to be thoroughly tested.

Pl217 committed:
Group rows to update by versionTags

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 committed:
Skip validation for `attachment` and `measurement` queries

Due to performance issues `updateVersionStates` has seen,
along with other methods it calls, we need to sacrifice
a little bit of data validation safety, which takes a
significant amount of time when a plan with lots of data
is being published.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

krishnareddy-tadi committed:
Add condition to ignore cluster cost attachments in plan publish

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review
